### ISSUE TITLE
Introduce a feature for bottom-up page-to-layout communication

### ISSUE THREAD

**Issue by aradalvand** [sveltejs/kit#5951]
### Describe the problem

Prior to the recent API overhaul (#5748), I used to have a `Meta.svelte` component that was rendered by my root layout. This component would read from `$page.stuff` — which I had typed in `app.d.ts` — and set various page-level head tags such as `<title>`, `<meta name="description">` and so on based on those `stuff` values. Every page would then have to provide this information in its `load` function, like so:
```ts
export function load() {
  return {
    stuff: {
      meta: {
        title: 'Home page',
        titleSuffix: 'long',
        description: 'Blah blah blah'
      }
    }
  };
}
```

My `app.d.ts` file looked something like:
```ts
declare namespace App {
  interface Stuff {
    meta: {
      title: string;
      titleSuffix?: 'short' | 'long' | 'none';
      description?: string;
    };
  }
}
```

And the `<Meta>` component:
```svelte
<script lang="ts">
  import { page } from '$app/stores';

  const titleSuffixes = new Map<TitleSuffix, string>([
    ['short', ' | Foo'],
    ['long', ' | Foo Bar Buzz'],
    ['none', '']
  ]);

  $: ({
    title,
    titleSuffix = 'short',
    description,
  } = $page.stuff.meta);
</script>

<svelte:head>
  <title>{title + titleSuffixes.get(titleSuffix)}</title>

  {#if description}
    <meta name="description" content={description} />
  {/if}
</svelte:head>
```

Now, ever since the API has been re-designed and the `stuff` thing has been removed, I've been struggling to re-create this properly.
Asking questions on the Discord server and reading [this part](https://github.com/sveltejs/kit/discussions/5774#discussioncomment-3292693:~:text=The%20data%20from%20all%20your%20load%20functions%20(or%20%2Blayout.server.js/%2Bpage.server.js%20files)%20is%20merged%20into%20a%20single%20object%20that%20can%20be%20used%20anywhere%20in%20your%20app%20via%20%24page.data%2C%20which%20replaces%20%24page.stuff.) of the migration guide lead me to think that returning this information as part of the data returned by each load function is apparently the way things like this are meant to be handled now. But I find that this is an almost unacceptable alternative, for two major reasons + a more minor one:
1. No type-safety while providing the values in the load functions:
```ts
export const load: PageLoad = () => {
  return {
    meta: {
      title: true, // Nothing's preventing me from doing this
      titleSuffix: 'somenonsense', // or this
      descriptititoitnotintotn: 'Blah blah blah', // or this 
    }
  };
};
```
2. No type-safety while reading the values in the `<Meta>` component

`Meta.svelte`:
```svelte
<script lang="ts">
    import { page } from '$app/stores';
</script>

<svelte:head>
    <!-- Absolutely no type-safety here, `meta` is `any` -->
    <title>{$page.data.meta.title}</title>
</svelte:head>
```
3. These values will be visible from the `+page.svelte` file included as part of its `data` prop, which is weird since it's actually none of its business.

![image](https://user-images.githubusercontent.com/26527405/184986989-6fc15e7c-b041-4c39-b2a6-330e676527a0.png)
`meta` isn't really something that this page is concerned with, so this is slightly awkward.

This (particularly the lack of type-safety) clearly yields a drastically poorer developer experience compared to the previous API. I sense that not enough thought has been put into this particular requirement when re-designing the API, namely bottom-up communication between pages and layouts. Page data isn't really fit for this use case at all.

### Describe the proposed solution

I don't have a clear-cut proposal here, the general idea is that perhaps you could consider introducing a new construct designed specifically for bottom-up page-to-layout communication, in order to properly accommodate use cases like this, which are by no means uncommon or unusual, or niche. It actually gets more complex in a real-world app where there are typically multiple levels of layouts, and each of them receives a different set of configurations/settings; which is why I believe this is something that deserves a dedicated feature, so to speak.

### Importance

crucial

### Additional Information

Please don't just hand wave this away, thanks. I know it's annoying that literally dozens of issues get created on this repo every single day and naturally it's almost impossible to give each and every single one of them a lot of attention, but this one is a real problem, I think, and one that is in need of a response from your side.

And tell me if I'm missing something, because I certainly raised an eyebrow while trying to do this but failing to find a satisfying solution, so it's possible that I'm unaware of something here.

**Comment by rmunn** [comment]
The discussion at https://github.com/sveltejs/kit/discussions/5748#discussioncomment-3276121 suggested using stores for this purpose. The top-level layout puts a store, let's say it's called `meta`, into `data`. Then it subscribes to that store to make any changes (to title, etc.) that are needed. The page then does something like this (note: code NOT tested, just written into the comment box, so it could well be wrong in significant details):

```js
const { meta } = data;
$meta.title = 'New title for this page';
```

That should fire the subscription in the top-level layout, which would then update the title with the new value.

**Comment by dummdidumm** [comment]
I don't think we will reintroduce something else, the "this pops up in my data" is negligible to me - you don't have to use it in your component. The type story is worse for now, but we will implement some level of type safety again for this.

**Comment by aradalvand** [comment]
@rmunn That's an interesting idea, but I tried it and it doesn't work during SSR which is a total deal breaker, as you would obviously need meta information like `<title>` to be set during SSR.

> That should fire the subscription in the top-level layout, which would then update the title with the new value.

Unfortuantely that's precisely what doesn't happen (during SSR), for some unfathomable reason, not sure if there's an existing issue for this particular problem?

**Comment by aradalvand** [comment]
> I don't think we will reintroduce something else, the "this pops up in my data" is negligible to me - you don't have to use it in your component. The type story is worse for now, but we will implement some level of type safety again for this.

@dummdidumm As I said, that one is the most minor issue, but I thought it's worth mentioning as it could be indicative of a larger conceptual problem. The type-safety is non-existent, and I'm not sure exactly why you're opposed to introducing a new distinct construct for this, but even if you don't want to do that, at the very least, [using stores](https://github.com/sveltejs/kit/issues/5951#issuecomment-1217302602) here should be possible in this context. As soon as I read @rmunn's comment I thought yeah that's probably the perfect solution for this, but it doesn't work during SSR and I'm not sure why, would appreciate it if you explain this.

**Comment by Rich-Harris** [comment]
The type safety you had before was illusory — nothing was forcing your page `load` functions to return the properties you needed on `stuff`. You had type safety insofar as it would prevent you from returning the _wrong_ properties, but it didn't guarantee the _right_ properties. Runtime checking was always necessary.

Stores won't work for this in SSR because you're not setting the value until the leaf is rendered, by which point the layout... has already been rendered. `$page.data` works because the data for the page is resolved before rendering happens.

**Comment by aradalvand** [comment]
@Rich-Harris So then what approach would you take for implementing things like this currently? And it's not just meta information, you may have a `hideFooter` "layout setting" that needs to be set by the pages, and things of that nature in general, and it gets worse if you have multiple level layouts.

**Comment by abdo643-HULK** [comment]
> @Rich-Harris So then what approach would you take for implementing things like this currently? And it's not just meta information, you may have a `hideFooter` "layout setting" that needs to be set by the pages, and things of that nature in general?
> 
> This gets worse if you need, for example, default values for these settings, all of this renders `$page.data` unsuitable for this use case; that's the point I'm trying to make.

A derived store in the root where you read the props you need from `$page.data`

**Comment by aradalvand** [comment]
@abdo643-HULK That still has all the downsides and limitations of `$page.data` that I've touched upon. Doesn't really solve anything.

**Comment by aradalvand** [comment]
Something like `$page.statics` proposed in #1459 (+ #269) is more or less what I'm suggesting. Those were deemed resolved when `stuff` was introduced apparently, but now we're back to where we were since there's no `stuff` anymore and the supposed alternative which is `$page.data` is not really fitting for such use cases.

**Comment by abdo643-HULK** [comment]
Because it is a derived store you can create your own type for ts. And you also said that things like meta aren't concerns of the `+page.svelte` but I don't agree with that because each page has it's own meta data. Just because the component doesn't consume it because of layouts doesn't make it not part of the page

**Comment by aradalvand** [comment]
> Because it is a derived store you can create your own type for ts.

You can create your own TS type already, and just convert `$page.data.meta` to it like `$page.data.meta as YourType`, so introducing a derived store is unnecessary.

> And you also said that things like meta aren't concerns of the +page.svelte but I don't agree with that because each page has it's own meta data. Just because the component doesn't consume it because of layouts doesn't make it not part of the page data

Perhaps this wasn't the best example, but I also provided other examples such as `hideFooter`, etc. The basic requirement is simple, a layout can have certain "settings" that need to be set by every page that derives from them, that's not really an exotic idea, so my point stands.

**Comment by aradalvand** [comment]
It is also awkward to have to create a separate `+page.js` file and a `load` function only to set some layout settings, even though the page doesn't actually do anything else in its `load` function. I think this all clearly begs for a separate feature.

To be honest I find it a little frustrating to have to argue for this as I think it should be obvious after minimal examination.

**Comment by dummdidumm** [comment]
You think you "elaborately" argue, I think you are being passive-aggressive. `stuff` was confusing to a lot of people, and `$page.data` achieves the same with a uniform solution which satisfied both the layout use case as well as the "don't want to do prop drilling" use case. What I decipher from your posts is that you don't want to mix layout concerns with other data in a single store, which doesn't sound like a concern to me. You are free to introduce a convention to your code that all layout-related data should be within a `layout` prop or sth like that. Your argument of needing to provide a `load` function was true for `stuff`, too, so I don't see a drawback here - and Rich explained why it isn't possible to do this in `+page.svelte`, because it's too late by then for SSR.

What's left is the TypeScript-story, which I agree needs to get better.

**Comment by aradalvand** [comment]
> Your argument of needing to provide a load function was true for stuff, too, so I don't see a drawback here

@dummdidumm Yes but it's worsened by virtue of the fact that now you need to have a separate file for the `load` function.

> What I decipher from your posts is that you don't want to mix layout concerns with other data in a single store, which doesn't sound like a concern to me.

It's deeper than that. Providing default values for example, as I mentioned above, would be non-obvious and awkward in this approach. And what if you have multiple levels of layouts? The [merging behavior](https://github.com/sveltejs/kit/discussions/5774#discussioncomment-3292693:~:text=merged%20into%20a%20single%20object) won't work as expected in that case, since it's a shallow merge. Consider the following example:

`routes/+layout.svelte`:
```svelte
<script>
  import { page } from '$app/stores';
</script>

<Header style={$page.data.layout.headerStyle} />

<!-- As a side note, this is more or less how you would have to provide default values, and it's not ideal since the default values are scattered around the entire file and you would potentially have to repeat them in multiple places or otherwise resort to more awkward workarounds -->
{#if $page.data.layout.showFooter ?? true}
  <Footer />
{/if}
```

`routes/whatever/+layout.js`:
```js
export function load(){ 
  return {
    layout: {
      headerStyle: 'secondary'
    }
  };
}
```

`routes/whatever/deeper/+page.js`:
```js
export function load(){ 
  return {
    layout: {
      showFooter: false,
    }
  };
}
```
The resulting `$page.data` here would not include the values that `routes/whatever/+layout.js` set. Since `routes/whatever/deeper/+page.js` essentially overrode the entire object even though it really meant to set a single property. Yes I know I can work around this like:
`routes/whatever/deeper/+page.js`:
```js
export async function load({ parent }){ 
  const { layout } = await parent();
  return {
    layout: {
      ...layout
      showFooter: false,
    }
  };
}
```
But now these two load functions are run serially as opposed to concurrently and that would have a performance cost if they're sending fetch requests and whatnot. I don't really *need* them to run serially, this is just what I *have* to do in this case because there's no other way to achieve what is needed. Hope that makes sense.

The same problem more or less existed for `stuff` too, it's not like I'm a huge fan of `stuff`, I'm just saying that this is the sort of thing that requires a native feature to specifically account for, because doing it using already-existing features (page data in this case) clearly yields a lot of awkwardness.

**Comment by pilcrowOnPaper** [comment]
Maybe introducing route specific hooks/handle could solve it without adding massive changes to the APIs? It seems like the problem is we don't having page-specific things that run before the whole rendering process. Having something like `route/+hooks.server.js`, @aradalvand use case could be solved like this:

```ts
// /some-page/+hooks.js

export const handle: Handle = async (event, resolve) => {
    event.locals.meta = {
        title: "random-title",
        includeFooter: true
    }
    return await resolve(event)
}
```

```ts
// /+layout.server.ts

export const load: PageLoad = async ({ locals }) => {
    return locals.meta
}
```

And the meta data can be accessed via `$page.data`.

**Comment by pilcrowOnPaper** [comment]
Maybe route specific hooks might partially address the problem (#5763), like `/some-route/+hooks.js`?

**Comment by tcc-sejohnson** [comment]
#6017 will mostly fix this issue, allowing `load`s to run in parallel but data resolution to occur as a (partial or complete) waterfall.

Automatic types for `$page.data` would be chaotic at best, completely unusable at worst, given that whatever type we gave it would have to be the _complete, merged result of every possible `load` return from every possible route, wrapped in `Partial`_. I mean, it's _technically_ possible, but at that point would essentially just be autocomplete.

**Comment by Rich-Harris** [comment]
One thought that occurs to me: in the same way we used to have `App.Stuff`, we could have an `App.PageData` type that _every_ `PageLoad` function is required to return, then we could use that same type for `$page.data`:

```ts
// src/app.d.ts
declare namespace App {
  interface PageData {
    title: string;
  }
}
```


```ts
// $types.d.ts
export type PageLoad = Kit.PageLoad<Params, InputData, App.PageData & Record<string, any>>;
```

```ts
import type { PageLoad } from './$types';

export const load: PageLoad = () => ({
  title: `if this is missing, it'll cause red squigglies`,
  other: {
    stuff: 'yay'
  }
});
```

```diff
// types/ambient.d.ts inside Kit
export interface Page<Params extends Record<string, string> = Record<string, string>> {
  url: URL;
  params: Params;
  routeId: string | null;
  status: number;
  error: HttpError | Error | null;
-  data: Record<string, any>;
+  data: App.PageData & Record<string, any>;
}
```

A fancier version would allow a `PageLoad` function _not_ to return `title` if a parent layout already did (I _think_ that's possible), so that a group of pages could share a `title`.

**Comment by tcc-sejohnson** [comment]
@Rich-Harris 

I'd make one change: `App.Data` wrapped in `Partial` and make it apply to Layout `load`s as well. Making it required for ALL `load`s seems extremely restrictive and would limit its usefulness. Wrapping it in `Partial` at least guarantees assignment type-safety but allows for more flexibility.

**Comment by tcc-sejohnson** [comment]
> It wouldn't apply to all `load` functions, just `PageLoad`. Making it `Partial` defeats the object, which is for typechecking to fail if a page's `load` function doesn't return the required data

I'm saying that the object shouldn't be "A page's load should return all of the required data", because the number of cases in which all of my pages share a set of data that they should all return 100% of the time is 0. What actually _does_ happen is that all of my pages share a set of data that they _might or might not_ return, and what's important is that if they _do_ return it, that it be under the correct key and have the correct type. `Partial<Page.Data>` solves a wider, more flexible problem, whereas `Page.PageData` solves a very specific, (probably?) uncommon scenario where you can guarantee that every page in your entire app should return the same set of values.

**Comment by aradalvand** [comment]
> because the number of cases in which all of my pages share a set of data that they should all return 100% of the time is 0. 

Although I get your point, that's not true, if you return page metadata like title, description, and so on as part of the data, then every page would return those values.
And if you don't have such a scenario, then why not just make the properties in `App.PageData` optional? Am I missing something?

**Comment by Rich-Harris** [comment]
Right — the goal here is to be able to guarantee that you can do this sort of thing:

```svelte
<svelte:head>
  <title>{$page.data.title}<title>
</svelte:head>
```

For that to work, every `PageLoad` function _must_ return `title`. Half-measures like `Partial` take us back to the illusory type safety we had with `App.Stuff`.

It sounds like you want an `App.DataThatMightExist` interface. I'm not personally convinced there's a good use case for it (particularly post-#5994), but if there is, I think it's a separate discussion (other than making sure we pick a future-proof name for `App.PageData`).

**Comment by tcc-sejohnson** [comment]
> > because the number of cases in which all of my pages share a set of data that they should all return 100% of the time is 0. 
> 
> That's not necessarily true, if you return page metadata like title, description, and so on as part of the data, then every page would return those values.
> And if you don't have such a scenario, then why not just make the properties in `App.PageData` optional yourself? Am I missing something?

That's true -- you could make the properties optional... would that allow us to use this for layout `load`s? My biggest concern here is that we create a solution that works for both.

**Comment by aradalvand** [comment]
`App.SharedPageData` may be a more accurate name than `App.PageData` by the way.

**Comment by Rich-Harris** [comment]
Or `App.RequiredPageData`/`App.RequiredData`

**Comment by aradalvand** [comment]
> Or App.RequiredPageData/App.RequiredData

`App.SharedData` may be a better option since not everything in that type is necessarily "required", some properties could be optional.

**Comment by aradalvand** [comment]
> Or App.RequiredPageData/App.RequiredData

`App.SharedPageData` is probably more accurate since not everything within that type is necessarily "required". Some properties could be optional.

**Comment by tcc-sejohnson** [comment]
> For that to work, every PageLoad function must return title. Half-measures like Partial take us back to the illusory type safety we had with App.Stuff.

This just isn't true. 

`Partial<{ foo: string }>` disallows assigning anything but a `string` or `undefined` to `foo`. It says "if `foo` exists, it is definitely a string", which eliminates a whole class of runtime type checks and allows me to just say `$page.data?.foo ?? 'fallback'` and never think about it again. Whereas without it, `$page.data.fou` is very typo-able and has the same type as `$page.data.foo` -- `any`.

**Comment by aradalvand** [comment]
What if different routes have different shared page data by the way? This is not unthinkable.
Imagine you have a `src/routes/panel`, and a `+layout.svelte` file in there, you might want every page within that sub-directory/route to have some special shared data (like `activeMenuItem` or something).

**Comment by tcc-sejohnson** [comment]
> What if different routes have different shared page data by the way? This is rather common, I believe.
> Imagine you have `src/routes/panel` with a `+layout.svelte` and some pages deep inside, you might want every page within that sub-directory/route to have some special shared data (like `activeMenuItem` or something).

This is exactly why I advocate for `Partial<App.Data>` and extending the definition to `LayoutLoad` as well. `App.Data` is just way too restrictive for real-world uses.

**Comment by aradalvand** [comment]
> This is exactly why I advocate for Partial<App.Data> and extending the definition to LayoutLoad as well. App.Data is just way too restrictive for real-world uses.

But that doesn't seem like the optimal solution, ideally we should have "route-scoped" shared data, if that makes sense, instead of an app-wide one defined in `app.d.ts`.

**Comment by tcc-sejohnson** [comment]
> > This is exactly why I advocate for Partial<App.Data> and extending the definition to LayoutLoad as well. App.Data is just way too restrictive for real-world uses.
> 
> But that doesn't seem like the optimal solution, ideally we should have "route-scoped" shared data, if that makes sense, instead of an app-wide one defined in `app.d.ts`.

In an ideal world I'd agree with you, but we have to assign. `$page.data` _one and only one_` type, as there's no way to scope its TypeScript types to a specific route. (Maybe there is some hacky, super-hilariously-complicated way, but it'd be a fragile bodge-upon-a-bodge that would be _just looking_ for a way to break...)

**Comment by aradalvand** [comment]
> In an ideal world I'd agree with you, but we have to assign $page.data one and only one` type, as there's no way to scope its TypeScript types to a specific route. (Maybe there is some hacky, super-hilariously-complicated way, but it'd be a fragile bodge-upon-a-bodge that would be just looking for a way to break...)

Well why not though? Couldn't SvelteKit utilize the same trick used for `./$types` here as well? Something like `./$stores` or something, I guess.

**Comment by tcc-sejohnson** [comment]
> > In an ideal world I'd agree with you, but we have to assign $page.data one and only one` type, as there's no way to scope its TypeScript types to a specific route. (Maybe there is some hacky, super-hilariously-complicated way, but it'd be a fragile bodge-upon-a-bodge that would be just looking for a way to break...)
> 
> Well why not though? Couldn't SvelteKit utilize the same trick used for `./$types` here as well? Something like `./$stores` or something, I guess.

Because `$page.data` is imported from one module to everywhere, and therefore has only one type definition.

**Comment by babichjacob** [comment]
> ```svelte
> <svelte:head>
>  <title>{$page.data.title}<title>
> </svelte:head>
> ```
> For that to work, every `PageLoad` function must return `title`.

But a layout (whether it's the default layout or a nested or reset one) could return it on behalf of its section of pages, right? It doesn't make sense for `title` in the example, but for some other shared property, it could (example: stylizing various sections of a website with a `color`, like I do for my school notes).

**Comment by aradalvand** [comment]
These problems are the reason why I've been trying to make the point that bottom-up communication between pages and layouts is most likely best handled by some new construct/feature, as opposed to page data.

**Comment by aradalvand** [comment]
These problems are the reason why I've been trying to make the point that bottom-up communication between pages and layouts is most likely best handled by some new construct/feature, as opposed to page data, which is more suitable for top-down stuff, and gets awkward when you try to use it for this type of scenario.

**Comment by aradalvand** [comment]
Correct me if I'm wrong but I think all we're trying to achieve here, fundamentally, is "set layout props" in a sense.
If a layout could declare regular props and a page could somehow set values for those props, this would solve everything, pretty much.

`+layout.svelte`:
```svelte
<script lang="ts">
  export let showFooter: boolean = true;
</script>
```

`+page.svelte`:
```svelte
<script lang="ts" context="module">
  import type { LayoutProps } from './$types';
  export layoutProps: LayoutProps = {
    showFooter: false,
  };
</script>
```

Tell me if I'm off about something.

**Comment by Rich-Harris** [comment]
> But a layout (whether it's the default layout or a nested or reset one) could return it on behalf of its section of pages, right? It doesn't make sense for `title` in the example, but for some other shared property, it could (example: stylizing various sections of a website with a `color`, like I do for my school notes).

Yep, that's what I meant by this:

> A fancier version would allow a `PageLoad` function _not_ to return `title` if a parent layout already did (I _think_ that's possible), so that a group of pages could share a `title`.

> best handled by some new construct/feature

The discussion won't get far without a concrete proposal. `$page.data` seems like a pretty decent solution, all things considered, especially if we add `App.PageData` (or whatever we call it).

> `Partial<{ foo: string }>` disallows assigning anything but a `string` or `undefined` to `foo`. It says "if `foo` exists, it is definitely a string", which eliminates a whole class of runtime type checks and allows me to just say `$page.data?.foo ?? 'fallback'` and never think about it again

Making `foo` optional seems like a decent solution to this:

```ts
declare namespace App {
  interface PageData {
    foo?: string;
  }
}
```

We could make it so that layout `load` functions return `Partial<App.PageData> & Record<string, any>` but page `load` functions return `App.PageData & Record<string, any>` (but can also get away with omitting properties that are set by parent `load` functions) and I think it would solve all the discussed use cases.

**Comment by tcc-sejohnson** [comment]
> We could make it so that layout load functions return Partial<App.PageData> & Record<string, any> but page load functions return App.PageData & Record<string, any> (but can also get away with omitting properties that are set by parent load functions) and I think it would solve all the discussed use cases.

That's an acceptable compromise for me. It covers the `layout` case and also the "required for all pages" case while still allowing explicitly optional "global" return properties for pages.

**Comment by madeleineostoja** [comment]
Just trying to parse through this issue, unless there's a PR imminently in the works for the lack of typing on `$page.data`, should just that request be split into a new issue so it's easier to track? It seems this issue is covering a lot of ground.

FWIW love the proposal of adding a `PageData` interface to the `App` namespace, that's how I was previously typing `Stuff`, which `$page.data` has taken the reigns for in my use case.

**Comment by Aaron2321d** [comment]
It seems to me that what we're trying to achieve here, fundamentally, is "set layout props", in a sense.
If you somehow make it so that every layout can define regular props and every page can then set those props, this will solve this whole domain of problems beautifully. And it also **makes it possible to define default values** (I agree with @aradalvand that with page data this will not be as easy)

`+layout.svelte`:
```svelte
<script lang="ts">
    export let showFooter: boolean = true;
</script>
```

`+page.ts`:
```ts
export const load: PageLoad = ({ setLayoutProps }) => {
    setLayoutProps({
        showFooter: true,
    });
};
```

This is just a raw idea, so the API could be refined.
@Rich-Harris Wouldn't something like this be feasible? Thank you.

**Comment by david-plugge** [comment]
I´m very unsure if this is even possible but i´l try to explain my idea:

The types pretty much travel down the tree including every layout until we reach the page we are currently working in.
If we could somehow manage to merge the `LayoutData` from all layouts before the current file into one type like this

```ts
// .sveltekit/types/src/routes/$types.d.ts
export type LayoutData = { session: { user: { id: string } } }

// .sveltekit/types/src/routes/profile/$types.d.ts
import type { LayoutData as ParentData } from '../$types';
export type LayoutData = { random: string } & ParentData;

// .sveltekit/types/src/routes/profile/settings/$types.d.ts
import type { LayoutData as ParentData } from '../$types';
export type PageData = { value: number } & ParentData
```

the result should be a merged type that includes all parent data types. Should also be possible with named layouts.

The problem would about how to actually use this type. One way could be importing the `page` store from `./$stores` instead of `$app/stores`.

Edit:
Maybe `./$stores` isnt a bad idea for param autocompletion within a route aswell?

**Comment by dummdidumm** [comment]
The problem with `./$types` (and likewise something like `./$stores`) is that it comes from a type root, you can't put something that is present at runtime in there, only types.

**Comment by david-plugge** [comment]
I´ve updated my comment.
Not sure if its possible to tell vite how to resolve relative paths

> The problem would about how to actually use this type. One way could be importing the `page` store from `./$stores` instead of `$app/stores` where vite resolves to the current `$app/stores` but typescript uses the types within `.svelte-kit/types`

**Comment by jimjimjimhaha** [comment]
Regarding the default value issue that a few people raised here, can't you do something like this?

`+layout.svelte`:
```svelte
<script lang="ts">
  $: ({
    showFooter = true,
  } = $page.data.layout);
</script>
```
